### ISSUE TITLE
ProgressiveMediaSource is resolving all subtitles unlike SingleSampleMediaSource

### ISSUE THREAD

**Issue by Blatzar** [google/ExoPlayer#9944]
TL;DR ProgressiveMediaSource consumes a lot of RAM and lags with many subtitles as all subtitles are downloaded unlike SingleSampleMediaSource. How to make it load on demand?

I want to use a specific subtitle decoder based on the contents of the subtitle file. To achieve this I used ProgressiveMediaSource and a custom SubtitleDecoder but when loading ~50 subtitles the memory usage is ~200mb higher than when using the same code and testcase with a SingleSampleMediaSource. This makes it load the video several seconds slower and the UI is unresponsive when resolving the subtitles. 

Looking at traffic from the app reveals that **when using a SingleSampleMediaSource only the preferred subtitle is fetched**, but when **using ProgressiveMediaSource all subtitles are fetched**. Is it possible to stop this?

No, the custom subtitle decoder is not at fault, when using `SubtitleDecoderFactory.DEFAULT.createDecoder(format)` the issue remains.

The issue was introduced in this commit:
https://github.com/LagradOst/CloudStream-3/commit/0849bffe4971bdff37bd64779b30c1f1e17a67a9

[Exoplayer is built with a MergingMediaSource ](https://github.com/LagradOst/CloudStream-3/blob/0849bffe4971bdff37bd64779b30c1f1e17a67a9/app/src/main/java/com/lagradost/cloudstream3/ui/player/CS3IPlayer.kt#L397-L399)

the `adb bugreport` will be sent with email as I don't want it public.

**Comment by icbaker** [comment]
If you see issues with `ProgressiveMediaSource` but everything works with `SingleSampleMediaSource`, why not just use `SingleSampleMediaSource`? `SingleSampleMediaSource` is currently the standard way to handle side-loaded subtitles in ExoPlayer. We are working towards using `ProgressiveMediaSource` for this (as part of a much larger project), but this functionality is currently gated behind the (deliberately) scarily named [`DefaultMediaSourceFactory#experimentalUseProgressiveMediaSourceForSubtitles`](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/source/DefaultMediaSourceFactory.html#experimentalUseProgressiveMediaSourceForSubtitles(boolean)).

Looking at your linked commit it seems like you're not using that method, and are instead manually constructing your `ProgressiveMediaSource`? If you want us to investigate the behaviour you're seeing with `ProgressiveMediaSource`, please provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) that demonstrates the problem in a way that we can build locally (ideally a fork of our demo app, so it's clear what customisations you've made). I'm afraid the commit you've linked to is not minimal enough.

If you just want to use a custom `SubtitleDecoder` (and are happy to use `SingleSampleMediaSource`) then you can do this by:
* Creating a custom `SubtitleDecoderFactory` that returns your custom `SubtitleDecoder` instance when appropriate.
* Creating a custom `RenderersFactory` that constructs a `TextRenderer` using this `SubtitleDecoderFactory`
* Passing this `RenderersFactory` to the `ExoPlayer.Builder` constructor.

**Comment by Blatzar** [comment]
Thank you for your reply! Using a custom `RenderersFactory` with `SingleSampleMediaSource` worked without issues! :people_hugging: 

> why not just use `SingleSampleMediaSource`?

I didn't know about `RenderersFactory` and thought a `ProgressiveMediaSource` was required for a custom SubtitleDecoder :(

Issue solved, but I can create a minimal reproducible example to debug the `ProgressiveMediaSource` if you'd want :+1:
